examples/fluids/src/setuplibceed.c

```diff
@@ -213,33 +210,230 @@ PetscErrorCode SetupBCQFunctions(Ceed ceed, PetscInt dim_sur, PetscInt num_comp_
+        CeedInt p_petsc = is_simplex ? (p_ceed * num_comp) : permutation_indices[p_ceed * num_comp];
+
+        interp[q * P + p_ceed] = basis_tabulation->T[0][((face * Q + q) * P * num_comp + p_petsc) * num_comp + c];
+        for (CeedInt d = 0; d < dim; d++) {
+          grad[(d * Q + q) * P + p_ceed] = basis_tabulation->T[1][(((face * Q + q) * P * num_comp + p_petsc) * num_comp + c) * dim + d];
+        }
+      }
+    }
+
+    // -- Cleanup
+    if (permutation) PetscCall(ISRestoreIndices(permutation, &permutation_indices));
+    PetscCall(ISDestroy(&permutation));
+  }
+
+  // Finally, create libCEED basis
+  CeedBasisCreateH1(ceed, elem_topo, num_comp, P, Q, interp, grad, q_points, q_weights, basis);
+  PetscCall(PetscFree(q_points));
+  PetscCall(PetscFree(interp));
+  PetscCall(PetscFree(grad));
+
+  PetscFunctionReturn(PETSC_SUCCESS);
+}
+
+// -----------------------------------------------------------------------------
+// Get CEED Basis from DMPlex
+// -----------------------------------------------------------------------------
+PetscErrorCode CreateBasisFromPlex(Ceed ceed, DM dm, DMLabel domain_label, CeedInt label_value, CeedInt height, CeedInt dm_field,
+                                   CeedQuadMode quadMode, CeedBasis *basis) {
+  PetscDS         ds;
+  PetscFE         fe;
+  PetscQuadrature quadrature;
+  PetscBool       is_simplex = PETSC_TRUE;
+  PetscInt        ds_field   = -1;
+
+  PetscFunctionBeginUser;
+
+  // Get element information
+  PetscCall(DMGetRegionDS(dm, domain_label, NULL, &ds, NULL));
+  PetscCall(DMFieldToDSField(dm, domain_label, dm_field, &ds_field));
+  PetscCall(PetscDSGetDiscretization(ds, ds_field, (PetscObject *)&fe));
+  PetscCall(PetscFEGetHeightSubspace(fe, height, &fe));
+  PetscCall(PetscFEGetQuadrature(fe, &quadrature));
+
+  // Check if simplex or tensor-product mesh
+  PetscCall(DMPlexIsSimplex(dm, &is_simplex));
+
+  // Build libCEED basis
+  if (is_simplex) {
+    PetscTabulation basis_tabulation;
+    PetscInt        num_derivatives = 1, face = 0;
+
+    PetscCall(PetscFEGetCellTabulation(fe, num_derivatives, &basis_tabulation));
+    PetscCall(BasisCreateFromTabulation(ceed, dm, domain_label, label_value, height, face, fe, basis_tabulation, quadrature, basis));
+  } else {
+    PetscDualSpace dual_space;
+    PetscInt       num_dual_basis_vectors;
+    PetscInt       dim, num_comp, P, Q;
+
+    PetscCall(PetscFEGetSpatialDimension(fe, &dim));
+    PetscCall(PetscFEGetNumComponents(fe, &num_comp));
+    PetscCall(PetscFEGetDualSpace(fe, &dual_space));
+    PetscCall(PetscDualSpaceGetDimension(dual_space, &num_dual_basis_vectors));
+    P = num_dual_basis_vectors / num_comp;
+    PetscCall(PetscQuadratureGetData(quadrature, NULL, NULL, &Q, NULL, NULL));
+
+    CeedInt P_1d = (CeedInt)round(pow(P, 1.0 / dim));
+    CeedInt Q_1d = (CeedInt)round(pow(Q, 1.0 / dim));
+
+    CeedBasisCreateTensorH1Lagrange(ceed, dim, num_comp, P_1d, Q_1d, quadMode, basis);
```

Comment:
We cut out the `quadMode` (convention would be `quad_mode`, right?) and this creation of our own `TensorH1Lagrange` basis if we could get the 1D basis matrices from the FE. Did we talk about how easy it would be to get that out of the FE @jedbrown?

Reply:
I've updated the `quad_mode` name. I'll close this for now.